File: benchmarks/merge_sort/merge_sort.rb
def merge(arr, left, mid, right, temp)
  i, j, k = left, mid + 1, left

  while i <= mid && j <= right
    if arr[i] <= arr[j]
      temp[k] = arr[i]
      i += 1
    else
      temp[k] = arr[j]
      j += 1
    end
    k += 1
  end

  while i <= mid
    temp[k] = arr[i]
    i += 1
    k += 1
  end

  while j <= right
    temp[k] = arr[j]
    j += 1
    k += 1
  end

  (left..right).each { |x| arr[x] = temp[x] }
end

def merge_sort(arr, left, right, temp)
  if left < right
    mid = left + (right - left) / 2
    merge_sort(arr, left, mid, temp)
    merge_sort(arr, mid + 1, right, temp)
    merge(arr, left, mid, right, temp)
  end
end

n = ARGV[0] ? ARGV[0].to_i : 100000

# Fill with descending values
arr = (1..n).to_a.reverse
temp = Array.new(n, 0)

merge_sort(arr, 0, n - 1, temp)

# Print checksum
puts arr.sum

File: benchmarks/merge_sort/merge_sort.js
function merge(arr, left, mid, right, temp) {
    let i = left, j = mid + 1, k = left;

    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }
    while (i <= mid) {
        temp[k++] = arr[i++];
    }
    while (j <= right) {
        temp[k++] = arr[j++];
    }
    for (let x = left; x <= right; x++) {
        arr[x] = temp[x];
    }
}

function mergeSort(arr, left, right, temp) {
    if (left < right) {
        const mid = left + Math.floor((right - left) / 2);
        mergeSort(arr, left, mid, temp);
        mergeSort(arr, mid + 1, right, temp);
        merge(arr, left, mid, right, temp);
    }
}

const n = parseInt(process.argv[2]) || 100000;

// Fill with descending values
const arr = [];
for (let i = n; i > 0; i--) {
    arr.push(i);
}
const temp = new Array(n).fill(0);

mergeSort(arr, 0, n - 1, temp);

// Print checksum
let sum = 0;
for (let i = 0; i < n; i++) {
    sum += arr[i];
}
console.log(sum);
